lib/ecto/schema.ex

```diff
@@ -1621,6 +1629,23 @@ defmodule Ecto.Schema do
     struct
   end
 
+  @spec escape_association_where(any, Macro.env) :: any
```

Comment:
We don't need to escape_association_where. Let's just invoke the code at compile time and store the where. Basically, we can remove this code and the keyword update. If somebody wants to lazily invoke a function, then they can use the MFA-tuple approach. This should simplify the code.

For the dynamic case, I will remove the anonymous functions it has inside.